Mathlib/Topology/Algebra/Order/Archimedean.lean

```diff
@@ -66,4 +67,24 @@ theorem dense_or_cyclic (S : AddSubgroup G) : Dense (S : Set G) ∨ ∃ a : G, S
   rcases h with ⟨ε, ε0, hε⟩
   exact cyclic_of_isolated_zero ε0 (disjoint_left.2 hε)
 
+variable [Nontrivial G] [DenselyOrdered G]
+
+/-- In a nontrivial densely linear ordered archimedean topological group,
+a subgroup is either dense or is cyclic, but not both.
+
+For a non-exclusive "or" version, see `AddSubgroup.dense_or_cyclic` above. -/
+theorem dense_xor'_cyclic (s : AddSubgroup G) :
+    Xor' (Dense (s : Set G)) (∃ a, s = .zmultiples a) := by
+  if hd : Dense (s : Set G) then
+    simp only [hd, xor_true]
+    rintro ⟨a, rfl⟩
+    exact not_denseRange_zsmul hd
+  else
+    simp only [hd, xor_false, id, zmultiples_eq_closure]
+    exact s.dense_or_cyclic.resolve_left hd
+
+theorem dense_iff_ne_zmultiples {s : AddSubgroup G} :
```

Comment:
Does it make sense to state this and the AddCircle results over *Group, and additivize?

Reply:
I'm adding multiplicative versions of all the dependencies in order to do this.

Reply:
See #22463